### PR TITLE
HHH-13788 Schema update try to recreate existing tables

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-13788

It should fix https://github.com/quarkusio/quarkus/issues/5883

### CODE REVIEW

File: hibernate-core/src/test/java/org/hibernate/test/schemaupdate/SchemaUpdateTestWithUseJdbcMetadataDeaultsTest.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+		new SchemaValidator().validate( metadata );
+
+		new SchemaUpdate().setHaltOnError( true )
+				.setOutputFile( output.getAbsolutePath() )
+				.setFormat( false )
+				.execute( EnumSet.of( TargetType.DATABASE, TargetType.SCRIPT ), metadata );
+
+		final String fileContent = new String( Files.readAllBytes( output.toPath() ) );
+		assertThat( "The update output file should be empty", fileContent, is( "" ) );
```
Comment (Sanne) [inline hibernate-core/src/test/java/org/hibernate/test/schemaupdate/SchemaUpdateTestWithUseJdbcMetadataDeaultsTest.java]: I'm not sure I understand this test. Why does it test that a schema update script is _empty_ ?

Shouldn't we rather test that the schema update is correctly applied?

BTW I tried to revert the fix, and the test will still pass - so there's something probably not working as expected?

File: hibernate-core/src/test/java/org/hibernate/test/schemaupdate/SchemaUpdateTestWithUseJdbcMetadataDeaultsTest.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+				.execute( EnumSet.of( TargetType.DATABASE ), metadata );
+
+		new SchemaValidator().validate( metadata );
+
+		new SchemaUpdate().setHaltOnError( true )
+				.setOutputFile( output.getAbsolutePath() )
+				.setFormat( false )
+				.execute( EnumSet.of( TargetType.DATABASE, TargetType.SCRIPT ), metadata );
+
+		final String fileContent = new String( Files.readAllBytes( output.toPath() ) );
```
Comment (Sanne) [inline hibernate-core/src/test/java/org/hibernate/test/schemaupdate/SchemaUpdateTestWithUseJdbcMetadataDeaultsTest.java]: should you use `getAbsolutePath` here as well?

File: hibernate-core/src/test/java/org/hibernate/test/schemaupdate/SchemaUpdateTestWithUseJdbcMetadataDeaultsTest.java (hunk 1 above)
Comment (dreab8) [reply hibernate-core/src/test/java/org/hibernate/test/schemaupdate/SchemaUpdateTestWithUseJdbcMetadataDeaultsTest.java]: > Shouldn't we rather test that the schema update is correctly? 

depends what you want to test, the issue is the that the update tries to recreate existing tables so we have to check that the update script is empty.

>  I tried to revert the fix, and the test will still pass  
You have to try it with MySql

File: hibernate-core/src/test/java/org/hibernate/test/schemaupdate/SchemaUpdateTestWithUseJdbcMetadataDeaultsTest.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+		final MetadataSources metadataSources = new MetadataSources( ssr );
+		metadataSources.addAnnotatedClass( TestEntity.class );
+
+		metadata = (MetadataImplementor) metadataSources.buildMetadata();
+		metadata.validate();
+	}
+
+	@After
+	public void tearsDown() {
```
[the remark this answers is not in the document]
Comment (dreab8) [reply hibernate-core/src/test/java/org/hibernate/test/schemaupdate/SchemaUpdateTestWithUseJdbcMetadataDeaultsTest.java]: a typo

File: hibernate-core/src/test/java/org/hibernate/test/schemaupdate/SchemaUpdateTestWithUseJdbcMetadataDeaultsTest.java (hunk 2 above)
Comment (dreab8) [reply hibernate-core/src/test/java/org/hibernate/test/schemaupdate/SchemaUpdateTestWithUseJdbcMetadataDeaultsTest.java]: `getAbsolutePath()` returns a `String` while `Files.readAllBytes` requires a `Path`

File: hibernate-core/src/test/java/org/hibernate/test/schemaupdate/SchemaUpdateTestWithUseJdbcMetadataDeaultsTest.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+import static org.junit.Assert.assertThat;
+
+/**
+ * @author Andrea Boriero
+ */
+@RunWith(Parameterized.class)
+public class SchemaUpdateTestWithUseJdbcMetadataDeaultsTest {
+
+	@Parameterized.Parameters
+	public static Collection<String> parameters() {
```
Comment (NathanQingyangXu) [inline hibernate-core/src/test/java/org/hibernate/test/schemaupdate/SchemaUpdateTestWithUseJdbcMetadataDeaultsTest.java]: btw, it seems you can use `String[]` directly

Comment (dreab8) [reply hibernate-core/src/test/java/org/hibernate/test/schemaupdate/SchemaUpdateTestWithUseJdbcMetadataDeaultsTest.java]: Thank @NathanQingyangXu I applied your suggestion

**Comment by dreab8** [comment]
@Sanne thanks for the review, I pushed a commit with some changes to the test and any further suggestion to improve it is more than welcome.

File: hibernate-core/src/test/java/org/hibernate/test/schemaupdate/SchemaUpdateTestWithUseJdbcMetadataDeaultsTest.java (hunk 1 above)
Comment (NathanQingyangXu) [reply hibernate-core/src/test/java/org/hibernate/test/schemaupdate/SchemaUpdateTestWithUseJdbcMetadataDeaultsTest.java]: should we specify mysql dialect for the testing, then?

File: hibernate-core/src/test/java/org/hibernate/test/schemaupdate/SchemaUpdateTestWithUseJdbcMetadataDeaultsTest.java (hunk 4 above)
Comment (NathanQingyangXu) [reply hibernate-core/src/test/java/org/hibernate/test/schemaupdate/SchemaUpdateTestWithUseJdbcMetadataDeaultsTest.java]: It should be me who say thank you. I borrowed the testing methodology in this PR and applied it in my PR: https://github.com/hibernate/hibernate-orm/pull/3144. During the process, I found we can use array directly. Glad I can help.

**Comment by Sanne** [comment]
hi @dreab8 , I'm back to trying this out and still have some issues.

When I run:

    ./gradlew :hibernate-core:test -Pdb=pgsql --tests SchemaUpdateWithUse*

The two new tests fail: `org.hibernate.tool.schema.spi.SchemaManagementException: Schema-validation: missing table [TEST_ENTITY]`

**Comment by dreab8** [comment]
Hi @Sanne, I checked and it seems the changes fix the issue for MySql but not for PostgreSQL. I'm investigating.

Thanks for testing with PostgreSQL.

**Comment by Sanne** [comment]
AFAIR this relates to the fact that we didn't have the right metadata - could possibly be solved now since we actually load it? Not sure if the timing is right in relation with schema gen.

But yes even if this is still not fixed, it might be easier to solve now.

**Comment by dreab8** [comment]
@yrodiere @Sanne  I locally rebased my branch on main and run the `SchemaUpdateWithUseJdbcMetadataDefaultsSettingToFalseTest` test with PostgreSQL and it passed, I pushed force... let's see if all the tests pass on Github :crossed_fingers:

**Comment by dreab8** [comment]
the SchemaUpdateWithUseJdbcMetadataDefaultsSettingToFalseTest  set `hibernate.temp.use_jdbc_metadata_defaults` to false so for the identifier the default UPPER case strategy is used, the but this is not correct for all dbs and so the schema validation process fails.

**Comment by dreab8** [comment]
@Sanne @yrodiere locally I have a fix for the default strategy issue, I need to test it a little more and not sure if it belongs to the same Jira  or if we need to create a separate Jira for it

**Pull request merged**